datastructure/binarysearchtree.py

```diff
@@ -0,0 +1,176 @@
+        else:  # Value > current_node.value
+            current_node.right = self.__delete(current_node.right, value)
+        return current_node
+
+    def __inorder_successor(self, current_node):
+        """Return a new node in a subtree that is greater than current_node (root) and
+            smaller than everything else"""
+        current_node = current_node.right
+        while current_node.left:
+            current_node = current_node.left
+        return current_node
+
+    def delete(self, value):
+        """Delete a node containing value in a tree"""
+        if self.root and self.find(value):
+            self.__delete(self.root, value)
+            self.count -= 1
+            return True
+        return False
+
+    def __preorder(self, current_node, item_list):
+        item_list.append(current_node.value)
+        if current_node.left:
+            self.__preorder(current_node.left, item_list)
+        if current_node.right:
+            self.__preorder(current_node.right, item_list)
+        return item_list
+
+    def preorder(self):
+        """Traverse Root - Left - Right"""
+        if self.root:
+            return self.__preorder(self.root, [])
+        else:
+            return []
+
+    def __postorder(self, current_node, item_list):
+        if current_node.left:
+            self.__postorder(current_node.left, item_list)
+        if current_node.right:
+            self.__postorder(current_node.right, item_list)
+        item_list.append(current_node.value)
+        return item_list
+
+    def postorder(self):
+        """Traverse Left - Right - Root"""
+        if self.root:
+            return self.__postorder(self.root, [])
+        else:
+            return []
+
+    def __inorder(self, current_node, item_list):
+        if current_node.left:
+            self.__inorder(current_node.left, item_list)
+        item_list.append(current_node.value)
+        if current_node.right:
+            self.__inorder(current_node.right, item_list)
+        return item_list
+
+    def inorder(self):
+        """Traverse Left - Root - Right"""
+        if self.root:
+            return self.__inorder(self.root, [])
+        else:
+            return []
+
+    def get_size(self):
+        return self.count
+
+    def __isvalid(self, current_node):
+        if not current_node:  # This is the base case
+            return True
+        elif current_node.left and current_node.value <= current_node.left.value:  # check if this is not a valid node
+            return False
+        elif current_node.right and current_node.value >= current_node.right.value:  # check if this is not a valid node
+            return False
+        return self.__isvalid(current_node.left) and self.__isvalid(current_node.right)  # This is the recursive case
+
+    def is_valid(self):
+        if self.root:
+            return self.__isvalid(self.root)
+        else:
+            return True
+
+    def min_value(self):
```

Comment:
did you finish these?

Reply:
these wont return the right value for most of these. I'll give you a hint, you can do this in a loop.

Reply:
fixed it